File: backend/modules/work_records/infrastructure/repositories/work_record_repository_impl.py
"""
Work Record Repository Implementation
Реализация репозитория через SQLAlchemy
"""

from typing import List, Optional
from sqlalchemy.orm import Session
from datetime import datetime

from modules.work_records.domain.entities.work_record import WorkRecord
from modules.work_records.infrastructure.database.models import WorkRecordModel


class WorkRecordRepository:
    """Реализация репозитория записей о работе"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def _to_entity(self, model: WorkRecordModel) -> WorkRecord:
        """Преобразовать модель БД в сущность"""
        return WorkRecord(
            id=model.id,
            mentor_id=model.mentor_id,
            date=model.date,
            hours=model.hours,
            description=model.description,
            category=model.category,
            status=model.status,
            notes=model.notes,
            created_at=model.created_at,
            updated_at=model.updated_at
        )
    
    def _to_model(self, entity: WorkRecord) -> WorkRecordModel:
        """Преобразовать сущность в модель БД"""
        return WorkRecordModel(
            id=entity.id,
            mentor_id=entity.mentor_id,
            date=entity.date,
            hours=entity.hours,
            description=entity.description,
            category=entity.category,
            status=entity.status,
            notes=entity.notes,
            created_at=entity.created_at,
            updated_at=entity.updated_at
        )
    
    def find_all(self) -> List[WorkRecord]:
        """Получить все записи"""
        models = self.db.query(WorkRecordModel).order_by(WorkRecordModel.date.desc()).all()
        return [self._to_entity(model) for model in models]
    
    def find_by_id(self, record_id: str) -> Optional[WorkRecord]:
        """Получить запись по ID"""
        model = self.db.query(WorkRecordModel).filter(WorkRecordModel.id == record_id).first()
        return self._to_entity(model) if model else None
    
    def find_by_mentor(self, mentor_id: str) -> List[WorkRecord]:
        """Получить записи по ментору"""
        models = self.db.query(WorkRecordModel)\
            .filter(WorkRecordModel.mentor_id == mentor_id)\
            .order_by(WorkRecordModel.date.desc())\
            .all()
        return [self._to_entity(model) for model in models]
    
    def find_by_date_range(self, start_date: str, end_date: str) -> List[WorkRecord]:
        """Получить записи за период"""
        models = self.db.query(WorkRecordModel)\
            .filter(WorkRecordModel.date >= start_date)\
            .filter(WorkRecordModel.date <= end_date)\
            .order_by(WorkRecordModel.date.desc())\
            .all()
        return [self._to_entity(model) for model in models]
    
    def create(self, work_record: WorkRecord) -> WorkRecord:
        """Создать запись"""
        model = self._to_model(work_record)
        self.db.add(model)
        self.db.commit()
        self.db.refresh(model)
        return self._to_entity(model)
    
    def update(self, work_record: WorkRecord) -> WorkRecord:
        """Обновить запись"""
        model = self.db.query(WorkRecordModel).filter(WorkRecordModel.id == work_record.id).first()
        if not model:
            raise ValueError(f"Work record with id {work_record.id} not found")
        
        model.date = work_record.date
        model.hours = work_record.hours
        model.description = work_record.description
        model.category = work_record.category
        model.status = work_record.status
        model.notes = work_record.notes
        model.updated_at = work_record.updated_at
        
        self.db.commit()
        self.db.refresh(model)
        return self._to_entity(model)
    
    def delete(self, record_id: str) -> bool:
        """Удалить запись"""
        model = self.db.query(WorkRecordModel).filter(WorkRecordModel.id == record_id).first()
        if not model:
            return False
        
        self.db.delete(model)
        self.db.commit()
        return True

File: frontend/modules/work-records/infrastructure/repositories/work-record.repository.impl.js
/**
 * Work Record Repository Implementation
 * Реализация работы с хранилищем записей о работе через Backend API
 */

import api from '../../../../core/shared/api.js';
import WorkRecordEntity from '../../domain/entities/work-record.entity.js';

class WorkRecordRepository {
  constructor() {
    this.endpoint = '/work-records';
  }

  /**
   * Преобразовать DTO из API в Entity
   */
  _dtoToEntity(dto) {
    return WorkRecordEntity.fromJSON({
      id: dto.id,
      mentorId: dto.mentorId,
      date: dto.date,
      hours: dto.hours,
      description: dto.description,
      category: dto.category,
      status: dto.status,
      notes: dto.notes,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt
    });
  }

  /**
   * Преобразовать Entity в DTO для API
   */
  _entityToDto(entity) {
    return {
      mentorId: entity.mentorId,
      date: entity.date,
      hours: entity.hours,
      description: entity.description,
      category: entity.category,
      status: entity.status,
      notes: entity.notes
    };
  }

  /**
   * Получить все записи
   */
  async findAll() {
    try {
      const data = await api.get(this.endpoint);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[WorkRecordRepository] Error fetching all:', error);
      throw new Error('Не удалось загрузить записи о работе');
    }
  }

  /**
   * Получить запись по ID
   */
  async findById(id) {
    try {
      const data = await api.get(`${this.endpoint}/${id}`);
      return this._dtoToEntity(data);
    } catch (error) {
      if (error.message.includes('404')) {
        return null;
      }
      console.error('[WorkRecordRepository] Error fetching by ID:', error);
      throw new Error('Не удалось загрузить запись');
    }
  }

  /**
   * Получить записи по ментору
   */
  async findByMentor(mentorId) {
    try {
      const data = await api.get(`${this.endpoint}?mentor_id=${mentorId}`);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[WorkRecordRepository] Error fetching by mentor:', error);
      throw new Error('Не удалось загрузить записи ментора');
    }
  }

  /**
   * Получить записи за период
   */
  async findByDateRange(startDate, endDate) {
    try {
      const data = await api.get(`${this.endpoint}?start_date=${startDate}&end_date=${endDate}`);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[WorkRecordRepository] Error fetching by date range:', error);
      throw new Error('Не удалось загрузить записи за период');
    }
  }

  /**
   * Получить записи за месяц
   */
  async findByMonth(year, month) {
    const records = await this.findAll();
    return records.filter(record => {
      const monthYear = record.getMonthYear();
      return monthYear.year === year && monthYear.month === month;
    });
  }

  /**
   * Сохранить запись
   */
  async save(workRecord) {
    try {
      const dto = this._entityToDto(workRecord);
      const data = await api.post(this.endpoint, dto);
      return this._dtoToEntity(data);
    } catch (error) {
      console.error('[WorkRecordRepository] Error saving:', error);
      throw new Error(error.message || 'Не удалось сохранить запись');
    }
  }

  /**
   * Обновить запись
   */
  async update(id, data) {
    try {
      const dto = this._entityToDto(data);
      const updated = await api.put(`${this.endpoint}/${id}`, dto);
      return this._dtoToEntity(updated);
    } catch (error) {
      console.error('[WorkRecordRepository] Error updating:', error);
      throw new Error(error.message || 'Не удалось обновить запись');
    }
  }

  /**
   * Удалить запись
   */
  async delete(id) {
    try {
      await api.delete(`${this.endpoint}/${id}`);
      return true;
    } catch (error) {
      if (error.message.includes('404')) {
        return false;
      }
      console.error('[WorkRecordRepository] Error deleting:', error);
      throw new Error('Не удалось удалить запись');
    }
  }

  /**
   * Получить статистику по ментору за период
   */
  async getMentorStats(mentorId, startDate, endDate) {
    const records = await this.findByMentor(mentorId);
    
    const filtered = records.filter(record => {
      if (!startDate && !endDate) return true;
      
      const recordDate = new Date(record.date);
      if (startDate && recordDate < new Date(startDate)) return false;
      if (endDate && recordDate > new Date(endDate)) return false;
      
      return true;
    });

    const totalHours = filtered.reduce((sum, record) => sum + record.hours, 0);
    const completedRecords = filtered.filter(r => r.isCompleted()).length;
    const approvedRecords = filtered.filter(r => r.isApproved()).length;

    return {
      totalRecords: filtered.length,
      totalHours,
      completedRecords,
      approvedRecords,
      records: filtered
    };
  }
}

export default WorkRecordRepository;

